spring-pulsar-spring-cloud-stream-binder/src/main/java/org/springframework/pulsar/spring/cloud/stream/binder/PulsarMessageChannelBinder.java

```diff
@@ -118,6 +129,40 @@ protected MessageProducer createConsumerEndpoint(ConsumerDestination destination
 		return pulsarMessageDrivenChannelAdapter;
 	}
 
+	// VisibleForTesting
+	@Nullable
+	Schema<Object> resolveSchema(@Nullable SchemaType schemaType, @Nullable Class<?> messageType,
+			@Nullable Class<?> messageKeyType, @Nullable Class<?> messageValueType) {
+		if (schemaType == null) {
+			schemaType = SchemaType.NONE;
+		}
+		ResolvableType resolvableType = null;
+		if (schemaType.isStruct()) {
+			resolvableType = ResolvableType.forClass(Objects.requireNonNull(messageType,
+					"'message-type' required for 'schema-type' " + schemaType.name()));
+		}
+		else if (schemaType == SchemaType.KEY_VALUE) {
+			resolvableType = ResolvableType.forClassWithGenerics(KeyValue.class,
+					Objects.requireNonNull(messageKeyType, "'message-key-type' required for 'schema-type' KEY_VALUE"),
+					Objects.requireNonNull(messageValueType,
+							"'message-value-type' required for 'schema-type' KEY_VALUE"));
+		}
+		else if (schemaType == SchemaType.NONE) {
+			if (messageType != null) {
+				resolvableType = ResolvableType.forClass(messageType);
+			}
+			else if (messageKeyType != null && messageValueType != null) {
+				resolvableType = ResolvableType.forClassWithGenerics(KeyValue.class, messageKeyType, messageValueType);
+			}
+			if (resolvableType == null) {
+				throw new IllegalArgumentException(
+						"'message-type' OR ('message-key-type' AND 'message-value-type') required for 'schema-type' NONE");
+			}
+		}
+		// TODO if schema == null then default lookup bean Schema<?> w/ name == binding
```

Comment:
This would be a nice fallback to add so that users can define a `Schema<?>` bean named `schemaFor<binding-name>` in case they need to provide a schema that we don't support w/ the above flow.